akka-cluster/src/main/scala/akka/cluster/MembershipState.scala

```diff
@@ -119,4 +124,137 @@ import akka.annotation.InternalApi
+          }
+
+        // chose another DC at random
+        val otherDcsInRandomOrder = dcsInRandomOrder((nodesPerDc - state.selfDc).keys.toList)
+        val unseenNodes = findFirstDcWithValidNodes(otherDcsInRandomOrder, onlyUnseen = true)
+        val unseenOrValidForGossip =
+          if (unseenNodes.nonEmpty) unseenNodes
+          else findFirstDcWithValidNodes(otherDcsInRandomOrder, onlyUnseen = false)
+
+        if (unseenOrValidForGossip.nonEmpty) unseenOrValidForGossip
+        // no other dc with reachable nodes, fall back to local gossip
+        else localDcGossipTargets(state)
+
+      }
+    }
+  }
+
+  private def oldestNMembersPerDc(state: MembershipState): Map[String, SortedSet[Member]] =
+    state.latestGossip.members.foldLeft(Map.empty[String, SortedSet[Member]]) { (acc, member) ⇒
+      acc.get(member.dataCenter) match {
+        case Some(set) if set.size > crossDcConnections ⇒ acc
+        case Some(set)                                  ⇒ acc + (member.dataCenter → (set + member))
+        case None                                       ⇒ acc + (member.dataCenter → (SortedSet.empty(Member.ageOrdering) + member))
+      }
+    }
+
+  /**
+   * For large clusters we should avoid shooting down individual
+   * nodes. Therefore the probability is reduced for large clusters.
+   */
+  protected def adjustedGossipDifferentViewProbability(clusterSize: Int): Double = {
+    val low = reduceGossipDifferentViewProbability
+    val high = low * 3
+    // start reduction when cluster is larger than configured ReduceGossipDifferentViewProbability
+    if (clusterSize <= low)
+      reduceGossipDifferentViewProbability
+    else {
+      // don't go lower than 1/10 of the configured GossipDifferentViewProbability
+      val minP = reduceGossipDifferentViewProbability / 10
+      if (clusterSize >= high)
+        minP
+      else {
+        // linear reduction of the probability with increasing number of nodes
+        // from ReduceGossipDifferentViewProbability at ReduceGossipDifferentViewProbability nodes
+        // to ReduceGossipDifferentViewProbability / 10 at ReduceGossipDifferentViewProbability * 3 nodes
+        // i.e. default from 0.8 at 400 nodes, to 0.08 at 1600 nodes
+        val k = (minP - reduceGossipDifferentViewProbability) / (high - low)
+        reduceGossipDifferentViewProbability + (clusterSize - low) * k
+      }
+    }
+  }
+
+  protected def selectDcLocalNodes: Boolean = ThreadLocalRandom.current.nextDouble() > 0.2D
+
+  protected def preferNodesWithDifferentView(state: MembershipState): Boolean =
+    ThreadLocalRandom.current.nextDouble() < adjustedGossipDifferentViewProbability(state.latestGossip.members.size)
+
+  protected def dcsInRandomOrder(dcs: List[DataCenter]): List[DataCenter] =
+    Random.shuffle(dcs)
+
+  protected def selectRandomNode(nodes: IndexedSeq[UniqueAddress]): Option[UniqueAddress] =
+    if (nodes.isEmpty) None
+    else Some(nodes(ThreadLocalRandom.current nextInt nodes.size))
```

Comment:
`ThreadLocalRandom.current.nextInt(nodes.size)`
(yeah, I know it was not written by you)